src/platform/intel/ace/include/ace/lib/mailbox.h

```diff
@@ -80,12 +80,12 @@ static inline uint64_t mailbox_sw_reg_read64(size_t offset)
 
 static inline void mailbox_sw_regs_write(size_t offset, const void *src, size_t bytes)
 {
-	int regs_write_err __unused = memcpy_s((void *)(MAILBOX_SW_REG_BASE + offset),
-					       MAILBOX_SW_REG_SIZE - offset, src, bytes);
+	uint32_t __sparse_cache *ptr_c;
+	volatile uint32_t *ptr;
 
-	assert(!regs_write_err);
-	dcache_writeback_region((__sparse_force void __sparse_cache *)(MAILBOX_SW_REG_BASE +
-								       offset), bytes);
+	ptr_c = (uint32_t __sparse_cache *)(MAILBOX_SW_REG_BASE + offset);
+	ptr = cache_to_uncache(ptr_c);
+	memcpy_s(ptr, MAILBOX_SW_REG_SIZE - offset, src, bytes);
```

Comment:
This is quite worrying finding, how come this has impact? Is the register read via cached address by other cores? This still doesn't explain why this PR would have any impact. We should reallly understand why dcache_writeback_region() is not working, we rely on this in many places.

Reply:
Yes, I am also very surprised. The reading method is mailbox_sw_reg_read, so it should be ok. This change use the same method for mailbox_sw_reg_read. 

And there is a PR to switch to uncached memory by who ?

Reply:
Linux driver doesn't access llp in mailbox now

Reply:
> dcache writeback is not enough, dcache_invalidate is needed on the reader core.

This would be a logical explanation, yes. Can we add such an invalidation on the reader side? But in principle - yes, a cached write plus a writeback shouldn't be faster than just a plain simple uncached write. We'd only benefit from cached access there if the same core were accessing that data multiple times without dropping caches - reading or writing (without writing back) data there. But I doubt that we ever do that.

Reply:
The reg reads seem to use uncached addresses, so these seem correct. Nevertheless, i don't think the writeback makes given the range of addresses may not be cache aligned, so best to change the register write routines to use uncached addresses.

Reply:
@lyakh these are "sw regs", so I guess it can be whatever sw/fw agree upon. Do you think the writeback is safe here?

Reply:
> @lyakh these are "sw regs", so I guess it can be whatever sw/fw agree upon. Do you think the writeback is safe here?

@kv2019i my understanding of "**software** registers" is that they're just generic locations in RAM used purely by the software without any direct hardware effects? If that's correct, then at the physical level they should also be handled like ordinary RAM, i.e. all cache management applies?

Reply:
My thoughs here are 
1) cache ops here are not being aligned or forced on cache line boundaries or sizes.
2) No obvious locking, so can race (and corrupt with 1 above).
3) The uncache alias is probably fixing this issue as the HW IO arbitration is probably enforcing serialised IO on the memory for all cores (i.e. the HW bus arbitration is a HW spinlock).

Reply:
> 3\. The



> My thoughs here are
> 
>     1. cache ops here are not being aligned or forced on cache line boundaries or sizes.

sw_regs_write may be no cache aligned since the llp slot size is 20 bytes

>     2. No obvious locking, so can race (and corrupt with 1 above).

we have lock to protect sw_regs_write

>     3. The uncache alias is probably fixing this issue as the HW IO arbitration is probably enforcing serialised IO on the memory for all cores (i.e. the HW bus arbitration is a HW spinlock).